### PR TITLE
remove docker and refactor

### PR DESCRIPTION
**Pull request by JanZachmann**

### CODE REVIEW

File: src/file/compression.rs [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                flate2::Compression::best(),
+            )),
+            Compression::xz => {
+                let range = 0..9;
+                let level = env::var("XZ_COMPRESSION_LEVEL")
+                    .unwrap_or_else(|_| "9".to_string())
+                    .parse()
+                    .unwrap_or(9);
+
+                let level = if range.contains(&level) { level } else { 9 };
```
Comment (empwilli) [inline src/file/compression.rs:54]: nit: maybe consider dropping line 48 for `if (0..9).contains(&level)`. Furthermore, while the result is the same, due to the else case, the range should be inclusive, i.e. `0..=9`.

File: src/file/functions.rs [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+        let v: Vec<&str> = s.split(&[',', ':']).collect();
+
+        anyhow::ensure!(v.len() == 3, err_msg);
+
+        let in_file = std::path::PathBuf::from(v[0]);
+        let partition = Partition::from_str(v[1])?;
+        let out_file = std::path::PathBuf::from(v[2]);
+
+        anyhow::ensure!(
+            in_file.try_exists().is_ok_and(|v| v),
```
Comment (empwilli) [inline src/file/functions.rs:75]: Why not `is_ok()`?

File: src/file/functions.rs [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+        } else {
+            tmp_out_file.push(param.out_file.file_name().unwrap());
+
+            let mut e2cp = Command::new("e2cp");
+            e2cp.arg(format!("{partition_file}:{in_file}"))
+                .arg(tmp_out_file.to_str().unwrap());
+            exec_cmd!(e2cp);
+            // since e2cp doesn't return errors in any case we check if output file exists
+            anyhow::ensure!(
+                tmp_out_file.try_exists().is_ok_and(|v| v),
```
Comment (empwilli) [inline src/file/functions.rs:357]: Why not `is_ok()`

File: src/lib.rs [hunk 4]
```diff
@@ -1,45 +1,99 @@
...
+    image_file: PathBuf,
+    generate_bmap: bool,
+    compress_image: Option<Compression>,
+    command: F,
+) -> Result<()>
+where
+    F: FnOnce(&PathBuf) -> Result<()>,
+{
+    anyhow::ensure!(
+        image_file.try_exists().is_ok_and(|v| v),
```
Comment (empwilli) [inline src/lib.rs:34]: Why not `is_ok()`?

File: src/ssh.rs [hunk 5]
```diff
@@ -54,7 +54,9 @@ impl Config {
 
         // check that key pair exists
         if let Some(key_path) = &priv_key_path {
-            if !key_path.exists() || !key_path.with_extension("pub").exists() {
+            if !key_path.try_exists().is_ok_and(|v| v)
+                || !key_path.with_extension("pub").try_exists().is_ok_and(|v| v)
```
Comment (empwilli) [inline src/ssh.rs:58]: Why not `is_ok()`?

File: src/ssh.rs [hunk 6]
```diff
@@ -263,7 +265,9 @@ pub async fn ssh_create_tunnel(
             (priv_key_path, pub_key_path)
         }
         Some(key_path) => {
-            if !key_path.exists() || !key_path.with_extension("pub").exists() {
+            if !key_path.try_exists().is_ok_and(|v| v)
+                || !key_path.with_extension("pub").try_exists().is_ok_and(|v| v)
```
Comment (empwilli) [inline src/ssh.rs:269]: Why not `is_ok()`?

**Review by empwilli — APPROVED** [review]
Minor nits.

File: src/file/functions.rs (hunk 2 above)
Comment (JanZachmann) [reply src/file/functions.rs:75]: it only returns the result but not if the file exists.
anyway I renamed v -> exists

File: README.md [hunk 7]
```diff
@@ -2,192 +2,128 @@
...
-Options:
-  -p <path>/payload.json
-  -b create bmap file
-```
-For further information on using dps payloads read the following [link](https://learn.microsoft.com/de-de/azure/iot-dps/concepts-custom-allocation).
+# Build from sources
 
-## Prepare `omnect-iotedge-gateway-device` and `omnect-iot-leaf-device` for a transparent gateway with leaf scenario
-Follow this article [Configure an IoT Edge device to act as a transparent gateway](https://docs.microsoft.com/en-us/azure/iot-edge/how-to-create-transparent-gateway?view=iotedge-2020-11) to understand the iotedge based transparent gateway setup. We assume that you use a X.509 CA certificate setup.
+The application can be built via `cargo` as usual. As a prerequisite `libmagic-dev` package must be installed on host system.
```
Comment (mlilien) [inline README.md:29]: if you build from sources, the dependency is libmagic ... 
libmagic-dev is a special package of debian based (host) buildsystems.

File: README.md [hunk 8]
```diff
@@ -2,192 +2,128 @@
...
-Copying files into or from the image is restricted to partions `boot`, `cert` and `factory`.
+### Copy files from image
 
-Note: If you need special permissions on copied files, you have to additionally copy a systemd-tmpfiles.d configuration file which handles these permissions.
-## Read e.g. `boot.scr`
+`omnect-cli` allows copying multiple files from multiple partitions in one command:
 
 ```sh
-omnect-cli file copy-from-image -f /boot.scr -i <path>/image.wic -p boot -d boot.scr
+omnect-cli file copy-from-image -f <partition>:<absolute-path-to-src>,<path-to-dest> -f <partition>:<absolute-path-to-src>,<path-to-dest> -i <path>/image.wic
```
Comment (mlilien) [inline README.md:87]: for me the format is not really intuitive.
e.g. i would expect something like `-f <absolute-path-to-src>,<partition>:<path-to-dest-in-partion>`

File: src/cli.rs [hunk 9]
```diff
@@ -157,36 +143,65 @@ pub enum IotHubDeviceUpdateConfig {
...
-    /// Linux).
-    #[arg(short = 'c', long = "config-path")]
-    pub config_path: Option<std::path::PathBuf>,
-    /// address of the backend API
-    #[arg(short = 'b', long = "backend", default_value = "https://cp.omnect.conplement.cloud")]
-    pub backend: String,
-    /// name of the device for which the ssh tunnel should be created.
-    pub device: String,
+/// file handling
+pub enum Ssh {
```
Comment (mlilien) [inline src/cli.rs:155]: Why not `SshConfig` all the other root structs are imho named *Config.
What is the difference here?

File: README.md (hunk 7 above)
Comment (mlilien) [reply README.md:29]: "A prerequisite is `libmagic`, e.g. the package `libmagic-dev` must be installed on a debian-based host system."
?

File: src/cli.rs (hunk 9 above)
Comment (JanZachmann) [reply src/cli.rs:155]: imho the `Connection` subcommand doesn't fit to `SshConfig`

Comment (mlilien) [reply src/cli.rs:155]: that creates a `ssh_config` file we use for connecting, e.g. via
` ssh -F /home/marcel/.local/share/omnect-cli/ssh_config test_device`

File: Cargo.toml [hunk 10]
```diff
@@ -1,58 +1,57 @@
...
 xz2 = "0.1"
 
 [dev-dependencies]
 httpmock = "0.6"
 file_diff = "1.0"
+assert_cmd = "2.0"
+
+# metadata for building with cargo-deb (https://crates.io/crates/cargo-deb) 
+[package.metadata.deb]
+depends = "bmap-tools, e2tools, fdisk, keychain, libc6 (>= 2.34), libmagic1, libssl3 (>= 3.0.0), mtools"
```
[the remark this answers is not in the document]
Comment (JanZachmann) [reply Cargo.toml:56]: should be a runtime dependency of omnect-cli:
https://github.com/omnect/omnect-cli/blob/27e5999d1544a3039657eeb958eadd51952ace00/src/ssh.rs#L80

File: src/cli.rs (hunk 9 above)
Comment (JanZachmann) [reply src/cli.rs:155]: Ok maybe `Ssh --> SshConfig` and `Connection --> SetConnection`?

File: testfiles/image.wic
Comment (mlilien) [inline testfiles/image.wic:1]: was this change intentional or was it accidentally created by a test?
if it was intentional, what changed?

Comment (JanZachmann) [reply testfiles/image.wic:1]: > was this change intentional or was it accidentally created by a test? if it was intentional, what changed?

yes it was. I had to add a line to /etc/hosts/ that can be replaced successfully by the regex

**Review by mlilien — APPROVED** [review]
